### PR TITLE
[CP-702] Add API for managing templates order

### PR DESCRIPTION
**Pull request by pawel-olejniczak**
**Description**

This API allows for managing message templates order.
Documentation: https://appnroll.atlassian.net/wiki/spaces/CP/pages/990642315/Messages+Endpoint+API+8#Change-a-message-template-order

**Your checklist for this pull request**

Make sure that this PR:
- [x] Complies with our guidelines for contributions
- [x] Has unit tests if possible.
- [x] Has documentation updated

### CODE REVIEW

**Review by kkleczkowski — APPROVED** [review]

File: image/user/db/sms_003.sql [hunk 1]
```diff
@@ -1,8 +1,8 @@
--- Copyright (c) 2017-2021, Mudita Sp. z.o.o. All rights reserved.
+-- Copyright (c) 2017-2022, Mudita Sp. z.o.o. All rights reserved.
 -- For licensing, see https://github.com/mudita/MuditaOS/LICENSE.md
 BEGIN TRANSACTION;
-INSERT OR REPLACE INTO "templates" ("_id","text","lastUsageTimestamp") VALUES (1,'Thanks for reaching out. I can''t talk right now, I''ll call you later',4);
-INSERT OR REPLACE INTO "templates" ("_id","text","lastUsageTimestamp") VALUES (2,'I''ll call you later',3);
-INSERT OR REPLACE INTO "templates" ("_id","text","lastUsageTimestamp") VALUES (3,'I''ll be there in 15 minutes',2);
-INSERT OR REPLACE INTO "templates" ("_id","text","lastUsageTimestamp") VALUES (4,'Give me 5 minutes',5);
```
Comment (SP2FET) [inline image/user/db/sms_003.sql:7]: won't this change damage DB in user's phones? afaik we don't have working database migraiton

Comment (pawel-mudita) [reply image/user/db/sms_003.sql:7]: Actually, it looks like it does break DB on phones with the newest OS version :(
This change is on hold until we decide how to deal with DB migration.

File: test/pytest/service-desktop/test_templates.py [hunk 2]
```diff
@@ -117,6 +116,37 @@ def test_changing_template_body(self):
         assert total_count == initial_count
         assert test_passed == True
 
+    def test_changing_template_order(self):
```
Comment (jurgenwigg) [inline test/pytest/service-desktop/test_templates.py:119]: Please provide documentation for this test: what is the scenario, the needed parameters for testing and a short description. Also please rework this test to not use `assert` in the middle of the test. If there is such a need please divide this test into smaller pieces or create a parametrization for this.

**Review by jurgenwigg — COMMENTED** [review]
Added class to `QAMuditaHarness` seems to be not used anywhere. If so, I suggest removing that change from this PR.

File: image/user/db/sms_003.sql (hunk 1 above)
Comment (lukasmas) [reply image/user/db/sms_003.sql]: As we have the UDM already done, the problem is fixed. 
I've added a new field according to the migration policy.

File: module-db/Interface/SMSTemplateRecord.hpp [hunk 3]
```diff
@@ -14,6 +14,7 @@ struct SMSTemplateRecord : public Record
 {
     UTF8 text;
     time_t lastUsageTimestamp = 0;
+    uint32_t order            = 0;
```
Comment (SP2FET) [inline module-db/Interface/SMSTemplateRecord.hpp:17]:
```suggestion
    std::uint32_t order            = 0;
```

File: module-db/Tables/SMSTemplateTable.hpp [hunk 4]
```diff
@@ -14,6 +14,7 @@ struct SMSTemplateTableRow : public Record
 {
     UTF8 text;
     time_t lastUsageTimestamp = 0;
+    uint32_t order            = 0;
```
[the remark this answers is not in the document]
Comment (lukasmas) [reply module-db/Tables/SMSTemplateTable.hpp]: Done.

File: products/PurePhone/services/db/databases/migration/sms/current/111a2d1b_Adding_new_column_to_templates_table_to_manage_templates_order/.meta [hunk 5]
```diff
@@ -0,0 +1,6 @@
+{
+ "id": "111a2d1b-dc32-40a3-9a2a-02a9e186bcac",
+ "date": "2023-03-23 10:11:44",
+ "message": "Adding new column to templates table to manage templates order",
+ "parent": 0
+}
```
Comment (rrandomsky) [inline products/PurePhone/services/db/databases/migration/sms/current/111a2d1b_Adding_new_column_to_templates_table_to_manage_templates_order/.meta:6]: No new line at the and of file

File: module-db/Tables/SMSTemplateTable.cpp [hunk 6]
```diff
@@ -21,10 +21,14 @@ bool SMSTemplateTable::create()
...
 {
-    return db->execute("INSERT or ignore INTO templates (text, lastUsageTimestamp) VALUES ('%q', '%q');",
-
-                       entry.text.c_str(),
-                       utils::to_string(entry.lastUsageTimestamp).c_str());
+    auto retQuery       = db->query("SELECT MAX(rowOrder) FROM templates;");
+    auto lastOrderValue = (*retQuery)[0].getUInt32() + 1;
+    return db->execute(
+        "INSERT or ignore INTO templates (text, lastUsageTimestamp, rowOrder) VALUES ('%q', '%q', '%lu');",
+
```
Comment (rrandomsky) [inline module-db/Tables/SMSTemplateTable.cpp:28]: remove line space

File: products/PurePhone/services/db/databases/migration/sms/current/111a2d1b_Adding_new_column_to_templates_table_to_manage_templates_order/.meta (hunk 5 above)
Comment (lukasmas) [reply products/PurePhone/services/db/databases/migration/sms/current/111a2d1b_Adding_new_column_to_templates_table_to_manage_templates_order/.meta]: Done

File: products/PurePhone/services/db/databases/migration/sms/current/111a2d1b_Adding_new_column_to_templates_table_to_manage_templates_order/up.sql [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+-- Message: Adding new column to templates table to manage templates order
+-- Revision: 111a2d1b-dc32-40a3-9a2a-02a9e186bcac
+-- Create Date: 2023-03-23 10:11:44
+
+ALTER TABLE templates
+ADD rowOrder INTEGER;
+
+UPDATE OR IGNORE templates
+SET rowOrder = _id;
+
```
[the remark this answers is not in the document]
Comment (lukasmas) [reply products/PurePhone/services/db/databases/migration/sms/current/111a2d1b_Adding_new_column_to_templates_table_to_manage_templates_order/up.sql]: Done

File: module-db/Tables/SMSTemplateTable.cpp (hunk 6 above)
Comment (lukasmas) [reply module-db/Tables/SMSTemplateTable.cpp]: Done

File: test/pytest/service-desktop/test_templates.py [hunk 8]
```diff
@@ -158,13 +188,23 @@ def test_get_templates_without_pagination(harness):
...
-def test_change_template(harness):
+def test_change_template_text(harness):
     templates_tester = TemplatesTester(harness)
     templates_tester.test_changing_template_body()
 
+
+@pytest.mark.rt1051
+@pytest.mark.service_desktop_test
+@pytest.mark.usefixtures("phone_unlocked")
+def test_change_template_order(harness):
```
[the remark this answers is not in the document]
Comment (lukasmas) [reply test/pytest/service-desktop/test_templates.py:203]: Same answer as above.

File: module-db/Interface/SMSTemplateRecord.cpp [hunk 9]
```diff
@@ -55,9 +56,11 @@ bool SMSTemplateRecordInterface::Update(const SMSTemplateRecord &rec)
     if (templ.ID == DB_ID_NONE) {
         return false;
     }
+    auto templateText  = rec.text.empty() ? templ.text : rec.text;
```
Comment (swichu91) [inline module-db/Interface/SMSTemplateRecord.cpp:59]:
```suggestion
    const auto templateText  = rec.text.empty() ? templ.text : rec.text;
```

File: module-db/Interface/SMSTemplateRecord.cpp [hunk 10]
```diff
@@ -55,9 +56,11 @@ bool SMSTemplateRecordInterface::Update(const SMSTemplateRecord &rec)
...
+    auto templateOrder = rec.order == 0 ? templ.order : rec.order;
```
Comment (swichu91) [inline module-db/Interface/SMSTemplateRecord.cpp:60]:
```suggestion
    const auto templateOrder = rec.order == 0 ? templ.order : rec.order;
```

File: module-db/Tables/SMSTemplateTable.cpp [hunk 11]
```diff
@@ -21,10 +21,13 @@ bool SMSTemplateTable::create()
 
 bool SMSTemplateTable::add(SMSTemplateTableRow entry)
 {
-    return db->execute("INSERT or ignore INTO templates (text, lastUsageTimestamp) VALUES ('%q', '%q');",
-
-                       entry.text.c_str(),
-                       utils::to_string(entry.lastUsageTimestamp).c_str());
+    auto retQuery       = db->query("SELECT MAX(rowOrder) FROM templates;");
```
Comment (swichu91) [inline module-db/Tables/SMSTemplateTable.cpp:24]: Return value check missing

File: module-db/Tables/SMSTemplateTable.cpp [hunk 12]
```diff
@@ -21,10 +21,13 @@ bool SMSTemplateTable::create()
...
+    auto lastOrderValue = (*retQuery)[0].getUInt32() + 1;
+    return db->execute(
+        "INSERT or ignore INTO templates (text, lastUsageTimestamp, rowOrder) VALUES ('%q', '%q', '%lu');",
```
Comment (swichu91) [inline module-db/Tables/SMSTemplateTable.cpp:27]: Note: You can also use macros from `module-db/Common/Types.hpp`

Comment (swichu91) [reply module-db/Tables/SMSTemplateTable.cpp:27]: Apart from that, replace `%lu` with either `u32_` or `PRIu32`

**Review by swichu91 — APPROVED** [review]

File: module-db/Tables/SMSTemplateTable.cpp (hunk 12 above)
Comment (lukasmas) [reply module-db/Tables/SMSTemplateTable.cpp:27]: Done

File: module-db/Tables/SMSTemplateTable.cpp (hunk 11 above)
Comment (lukasmas) [reply module-db/Tables/SMSTemplateTable.cpp]: Done

File: module-db/Interface/SMSTemplateRecord.cpp (hunk 10 above)
Comment (lukasmas) [reply module-db/Interface/SMSTemplateRecord.cpp]: Done

**Pull request merged**